db/commitlog/commitlog.cc

```diff
@@ -1675,6 +1679,71 @@ void db::commitlog::segment_manager::flush_segments(uint64_t size_to_remove) {
     totals.bytes_flush_requested += flushing;
 }
 
+void db::commitlog::segment_manager::check_no_data_older_than_allowed() {
+    if (!cfg.commitlog_data_max_lifetime_in_seconds) {
+        return;
+    }
+
+    auto now = gc_clock::now();
+
+    std::unordered_set<cf_id_type> ids;
+    std::optional<replay_position> high;
+
+    for (auto& s : _segments) {
+        auto rp = replay_position(s->_desc.id, db::position_type(s->size_on_disk()));
+        if (rp <= _flush_position) {
+            // already requested.
+            continue;
+        }
+
+        for (auto [id, low_ts] : s->_cf_min_time) {
+            // Ignore CF:s already released.
+            if (!s->_cf_dirty.count(id)) {
+                continue;
+            }
+
+            auto time_since_first_added = std::chrono::duration_cast<std::chrono::seconds>(now - low_ts).count();
+            if (time_since_first_added >= *cfg.commitlog_data_max_lifetime_in_seconds) {
+                // There is data in this segment that has lived longer than allowed (might be flushed actually
+                // but can still affect compaction/resurrect stuff on replay). Collect all dirty 
+                // id:s (stuff keeping this segment alive), and ask for them to be memtable flushed
+                for (auto& id : s->_cf_dirty | boost::adaptors::map_keys) {
+                    ids.insert(id);
+                }
+
+                // highest position (so far) is end of segment.
+                high = rp;
+
+                // If this segment is actually the active one, we must close it. Otherwise flushing 
+                // won't help.
+                if (s->is_still_allocating()) {
+                    // fixme. discarded future.
+                    (void)s->close();
+                }
+                continue;
+            }
+        }
+
+        // segments are sequential. Can't start becoming older.
+        break;
+    }
+
+    if (!ids.empty()) {
+        auto callbacks = boost::copy_range<std::vector<flush_handler>>(_flush_handlers | boost::adaptors::map_values);
+        // For each CF id: for each callback c: call c(id, high)
+        for (auto& f : callbacks) {
+            for (auto& id : ids) {
+                try {
+                    f(id, *high);
+                } catch (...) {
+                    clogger.error("Exception during flush request {}/{}: {}", id, high, std::current_exception());
```

Comment:
looks like warning is more appropriate than error, and please add test saying it's ignored.

Reply:
I think it is an error (and is an error on other flush paths). It should never happen, the path is expected to be noexcept, but it is a callback, so we remain vigilant.

Reply:
memtable flush is retryable (and it has its own error handling and logging).
Should we set `_flush_position = *high` after a failed flush?
Why do we even need to set it here?  I'd expect it to be updated via the normal flush path, and this code to only trigger stuff without changing the flush business logic.

Reply:
This is not waiting for flush or anything. This is telling memtables that we _want_ a flush, up to this RP. It is expected to return immediately, and be resolved eventually by memtables calling back with `discard_completed_segments`. 
This is equivalent (in contract) to flipping a bit and waking a cond or whatnot. 
`_flush_position` should perhaps be more aptly named `_flush_request_position`, it is the level we've currently requested to.

Reply:
We never request an RP range twice. I.e. "high" is always growing.

Reply:
I see.  worth a bit of verbiage to help novices like me :)